docs/ConstPropagationPass.md

```diff
@@ -0,0 +1,266 @@
+# Constant Propagation for ONNX operations
+
+This document describes `--constprop-onnx` pass which is used to do
+constant propagation for operations in the ONNX dialect.
+
+[source
+code](https://github.com/onnx/onnx-mlir/blob/master/src/Transform/ONNX/ConstProp.td).
+
+## Example
+Given the following code:
+```mlir
+func @foo() -> tensor<1xf32> {
+  %0 = "onnx.Constant"() {value = dense<[1.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %1 = "onnx.Constant"() {value = dense<[2.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %2 = "onnx.Add"(%0, %1) : (tensor<1xf32> , tensor<1xf32>) -> tensor<1xf32>
+  %3 = "onnx.Constant"() {value = dense<[3.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  %4 = "onnx.Add"(%2, %3) : (tensor<1xf32> , tensor<1xf32>) -> tensor<1xf32>
+  "std.return"(%4) : (tensor<1xf32>) -> ()
+}
+```
+
+If we call `onnx-mlir-op --constprop-onnx`, we will get:
+```mlir
+func @foo() -> tensor<1xf32> {
+  %0 = "onnx.Constant"() {value = dense<[6.0]> : tensor<1xf32>} : () -> tensor<1xf32>
+  "std.return"(%0) : (tensor<1xf32>) -> ()
+}
+```
+
+## Remark 
+
+ONNXConstantOp uses MLIR DenseElementsAttr to store constant values. It is
+important to note that, once a DenseElementsAttr is created, it is alive and 
+consumes memory until the end of compilation. In [Example](#example), all the
+three DenseElementsAttrs in the three ONNXConstantOps exist until the end of
+compilation. Especially, two intermediate DenseElementsAttrs in the two
+ONNXConstantOps produced by folding the two ONNXAddOps also exist. For a
+practice model, the number of intermediate DenseElementsAttrs will increase
+quickly, which lead to a large memory footprint during compilation. 
+
+To avoid creating too many DenseElementsAttrs for intermediate ONNXConstant ops
+during `--constprop-onnx`, we design a mechanism that dynamically allocates and
+deallocates buffers for intermediate ONNXConstantOps and only creates
+DenseElementsAttr for the final results of constant propagation.
+
+In particular, we maintain a buffer pool internally. When an ONNXConstantOp
+is reached at the first time, we read its DenseElementsAttr and store
+data to an array buffer in the pool. A unique buffer ID is used to map the
+ONNXConstantOp to the buffer. All constant computations are then done on array
+buffers, which is to avoid creating DenseElementsAttr for intermediate
+ONNXConstantOps. Buffers are automatically freed if they are not used.
+
+We provide three helper functions to use when working with buffers:
```

Comment:
1 seems to translate an array attribute into a buffer.
3 seems to create a new buffer
2 seems to associate a created buffer with an op.

I would reorder 2 & 3 as it seems more logical progression. What happen if we create a new buffer but don't associated it with an op? Does it get lost? (see request to have a bit more info for how the freeing of the buffers are done).

Is there ever a need to have an explicit freeing of a buffer? E.g. one that is never associated with an op?

Reply:
Thanks! It is better to reorder. Done.

> Is there ever a need to have an explicit freeing of a buffer? E.g. one that is never associated with an op?

Good question. I added an explanation for this situation. Yes, users need to take care of it and free it explicitly.

Reply:
tx